contract/src/main/kotlin/io/provenance/scope/loan/contracts/RecordLoanContract.kt

```diff
@@ -0,0 +1,94 @@
+
+    @Function(invokedBy = PartyType.OWNER)
+    @Record(LoanScopeFacts.asset)
+    open fun recordAsset(@Input(LoanScopeFacts.asset) asset: Asset) = asset.also {
+        // TODO: Add or implement correct extension for safe casting of kvMap values like "loan"
+        /*validateRequirements {
+            if (existingAsset != null) {
+                requireThat(
+                    // Flag that the asset is an eNote
+                    existingAsset.kvMap["loan"].isENote.isFalse() orError "asset cannot be updated",
+                    // optional: make sure nothing important changed
+                    // examples:
+                    existingAsset.id == asset.id orError "cannot change asset ID",
+                    existingAsset.type == asset.type orError "cannot change asset type",
+                    existingAsset.kvMap["loan"].originatorUuid == asset.kvMap["loan"].originatorUuid orError "cannot change loan originator UUID",
+                    existingAsset.kvMap["loan"].originatorName == asset.kvMap["loan"].originatorName orError "cannot change loan originator name",
+                )
+            } else {
+                requireThat(
+                    // other validation rules, such as:
+                    asset.id.isValid() orError "asset.id is missing",
+                    asset.type.isNotBlank() orError "asset.type is missing",
+                    asset.kvMap["loan"].originatorUuid.isValid() orError "asset.kv.loan.originatorUuid is missing",
+                    asset.kvMap["loan"].originatorName.isNotBlank() orError "asset.kv.loan.originatorName is missing",
+                )
+            }
+        }*/
+    }
+
+    @Function(invokedBy = PartyType.OWNER)
+    @Record(LoanScopeFacts.servicingRights)
+    open fun recordServicingRights(@Input(LoanScopeFacts.servicingRights) servicingRights: ServicingRights) = servicingRights
+
+    @Function(invokedBy = PartyType.OWNER)
+    @Record(LoanScopeFacts.documents)
+    open fun recordDocuments(@Input(LoanScopeFacts.documents) documents: LoanDocuments) = documents
+
+    @Function(invokedBy = PartyType.OWNER)
+    @Record(LoanScopeFacts.loanStates)
+    open fun recordLoanStates(@Input(LoanScopeFacts.loanStates) loanStates: LoanStates) = loanStates
+
+    @Function(invokedBy = PartyType.OWNER)
+    @Record(LoanScopeFacts.validationResults)
+    open fun recordValidationResults(@Input(LoanScopeFacts.validationResults) validationResults: ValidationResults) = validationResults
+
+    @Function(invokedBy = PartyType.OWNER)
+    @Record(LoanScopeFacts.eNote)
+    open fun recordENote(@Input(LoanScopeFacts.eNote) eNote: ENote) = eNote.also { // TODO: Confirm if we want existing and/or input to be nullable and adjust code below accordingly
+        validateRequirements {
+            if (existingENote != null) {
+                //requireThat(existingENote.checksum == eNote.checksum orError "cannot modify or remove eNote during loan onboarding"), // use specific contract instead
```

Comment:
Commented out for until checksum comparison is implemented or existing trailing comment is addressed